lib/puppet/util/servicenow.rb

```diff
@@ -95,42 +95,53 @@ def do_snow_request(uri, http_verb, body, user: nil, password: nil, oauth_token:
-      # corrective. Puppets API does not currently give us a way to distinguish
-      # between noop changes that would have been either corrective or
-      # intentional.
-      Puppet.info(sn_log_entry('decision: reportable pending_changes'))
-      return true
-    end
-
-    no_changes = status == 'unchanged' && !noop_pending
-
-    if no_changes && incident_creation_conditions.include?('no_changes')
-      Puppet.info(sn_log_entry('decision: reportable no_changes'))
-      return true
+  def create_incident?(report_status, resource_statuses, incident_creation_conditions)
+    # Some incident creation conditions depend on the resource events. Thus, we go ahead
+    # and evaluate these 'event' conditions _before_ evaluating the incident creation
+    # conditions so we can iterate through the resource events only once. This results in
+    # cleaner code for a negligible performance hit, where the performance hit occurs if the
+    # incident creation conditions do _not_ contain an 'event' condition.
+    #
+    # Note that event_conditions is a hash of <condition> => <satisfied?>
+    event_conditions = {
+      'corrective_changes'          => false,
+      'intentional_changes'         => false,
+      'pending_corrective_changes'  => false,
+      'pending_intentional_changes' => false,
+    }
+    resource_statuses.each do |_, resource|
+      resource.events.each do |event|
+        next if event.status == 'failure' || event.status == 'audit'
+        # event.status == 'success' || 'noop'. Either way, we found a satisfying
+        # change condition so determine its name
+        change_condition = event.corrective_change ? 'corrective_changes' : 'intentional_changes'
+        if event.status == 'noop'
+          change_condition = "pending_#{change_condition}"
+        end
+        event_conditions[change_condition] = true
+      end
     end
 
-    reportable_change = false
-
-    if corrective_change && incident_creation_conditions.include?('corrective_changes')
-      Puppet.info(sn_log_entry('decision: reportable corrective_changes'))
-      reportable_change = true
-    elsif status == 'changed' && !corrective_change && incident_creation_conditions.include?('intentional_changes')
-      Puppet.info(sn_log_entry('decision: reportable intentional_changes'))
-      reportable_change = true
+    # Now evaluate the satisfied conditions
+    satisfied_conditions = incident_creation_conditions.select do |condition|
+      if condition == 'always'
+        true
+      elsif condition == 'never'
+        false
+      elsif condition == 'failures'
+        report_status == 'failed'
+      elsif event_conditions.key?(condition)
+        event_conditions[condition]
+      else
+        # We should never hit this code-path
+        Puppet.warning(sn_log_entry("unknown incident creation conditon: #{condition}"))
+        false
+      end
     end
+    Puppet.info(sn_log_entry("satisfied conditions: #{satisfied_conditions}"))
```

Comment:
Biggest reason for this change (print _all_ satisfied conditions vs. print the first one that's true) is it felt weird to me that the latter prioritized certain conditions over others. For example, if I had a `failed` report with `pending` changes, why do the logs show that the failure's the reportable change when the pending change is also reportable?

Reply:
I think structuring it this way will also make https://tickets.puppetlabs.com/browse/PIE-311 a bit easier b/c the generated report labels will depend on the satisfied conditions. So we will need to know all of the latter anyways.